kafka/src/main/java/nl/nn/adapterframework/extensions/kafka/KafkaListener.java

```diff
@@ -0,0 +1,142 @@
+		if(patternRecheckInterval < 10) throw new ConfigurationException("patternRecheckInterval should be at least 10");
+		properties.setProperty(ConsumerConfig.GROUP_ID_CONFIG, groupId);
+		properties.setProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, fromBeginning ? "earliest" : "latest");
+		properties.setProperty(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
+		properties.setProperty(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName());
+		properties.setProperty(ConsumerConfig.METADATA_MAX_AGE_CONFIG, String.valueOf(patternRecheckInterval));
+		consumer = new KafkaConsumer<>(properties);
+	}
+
+	@Override
+	public void open() throws ListenerException {
+		Arrays.stream(topics.split(","))
+				.filter(StringUtils::isNotEmpty) //if topics is `abc` java will split it into `abc` and ``. The latter is not a valid topic.
+				.map(Pattern::compile) //Convert the topic to a regex pattern, to allow for wildcards in topic names.
+				.forEach(consumer::subscribe);
+	}
+
+	@Override
+	public void close() throws ListenerException {
+		consumer.close();
+	}
+
+	@Override
+	public Message extractMessage(
+			@Nonnull RawMessageWrapper<ConsumerRecord<String, byte[]>> wrappedMessage, @Nonnull Map<String, Object> context) throws ListenerException {
+		Map<String, String> headers=new HashMap<>();
+		ConsumerRecord<String, byte[]> rawMessage = wrappedMessage.getRawMessage();
+		Arrays.stream(rawMessage.headers().toArray()).forEach(header -> {
+			try {
+				headers.put(header.key(), new String(header.value(), StandardCharsets.UTF_8));
+			} catch(Exception e) {
+				log.warn("Failed to convert header key [{}] to string. Bytearray value: [{}]", header.key(), header.value(), e);
+			}
+		});
+		context.put("kafkaTopic", rawMessage.topic());
+		context.put("kafkaKey", rawMessage.key());
+		context.put("kafkaPartition", rawMessage.partition());
+		context.put("kafkaOffset", rawMessage.offset());
+		context.put("kafkaTimestamp", rawMessage.timestamp());
+		context.put("kafkaHeaders", headers);
+		return new Message(rawMessage.value(), new MessageContext(context));
+	}
+
+	@Override
+	public void afterMessageProcessed(PipeLineResult processResult, RawMessageWrapper<ConsumerRecord<String, byte[]>> rawMessage, PipeLineSession pipeLineSession) throws ListenerException {
+		// no-op
+	}
+
+	@Override
+	public String getPhysicalDestinationName() {
+		return "TOPICS(" + topics + ") on ("+ getBootstrapServers() +")";
+	}
+
+	@Nonnull
+	@Override
+	public Map<String, Object> openThread() throws ListenerException {
+		return new HashMap<>();
+	}
+
+	@Override
+	public void closeThread(@Nonnull Map<String, Object> threadContext) throws ListenerException {
+		// nothing.
+	}
+
+	@Override
+	public RawMessageWrapper<ConsumerRecord<String, byte[]>> getRawMessage(@Nonnull Map<String, Object> threadContext) throws ListenerException {
```

Comment:
What is the messages on the topic do not have a String Key, and Value of of `byte[]`?

Can it also be a stream of bytes or characters?

Could it be just a string?
And if it is one of the above then how should we handle that in client code?

I don't know how this works in Kafka.